src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -0,0 +1,763 @@
+//      } else {
+//        Fail("Operation not supported on the given types");
+//      }
+//    });
+//  } else if (right_is_null) {
+//    resolve_data_type(left_operand.data_type(), [&](const auto left_data_type_t) {
+//      using LeftDataType = typename decltype(left_data_type_t)::type;
+//      const auto left_result = evaluate_expression<LeftDataType>(left_operand);
+//      using ConcreteFunctor = Functor<T, LeftDataType, T>;
+//
+//      if constexpr (ConcreteFunctor::supported) {
+//        result = evaluate_binary_operator<T, LeftDataType, T>(left_result, ExpressionResult<T>(NullValue{}), ConcreteFunctor{});
+//      } else {
+//        Fail("Operation not supported on the given types");
+//      }
+//    });
+//
+//  } else {
+//    resolve_data_type(left_operand.data_type(), [&](const auto left_data_type_t) {
+//      using LeftDataType = typename decltype(left_data_type_t)::type;
+//
+//      const auto left_operands = evaluate_expression<LeftDataType>(left_operand);
+//
+//      resolve_data_type(right_operand.data_type(), [&](const auto right_data_type_t) {
+//        using RightDataType = typename decltype(right_data_type_t)::type;
+//
+//        const auto right_operands = evaluate_expression<RightDataType>(right_operand);
+//
+//        using ConcreteFunctor = Functor<T, LeftDataType, RightDataType>;
+//
+//        if constexpr (ConcreteFunctor::supported) {
+//          result = evaluate_binary_operator<T, LeftDataType, RightDataType>(left_operands, right_operands, ConcreteFunctor{});
+//        } else {
+//          Fail("Operation not supported on the given types");
+//        }
+//      });
+//    });
+//  }
+//
+//  return result;
+//}
+
+template<typename T>
+ExpressionResult<T> ExpressionEvaluator::evaluate_logical_expression(const LogicalExpression& expression) {
+  Fail("LogicalExpression can only output int32_t");
+}
+//
+//template<typename T>
+//ExpressionResult<T> ExpressionEvaluator::evaluate_in_expression(const InExpression& expression) {
+//  Fail("InExpression supports only int32_t as result");
+//}
+
+template<typename T>
+ExpressionResult<T> ExpressionEvaluator::evaluate_case_expression(const CaseExpression& case_expression) {
+  const auto when = evaluate_expression<int32_t>(*case_expression.when());
+
+  /**
+   * Handle cases where the CASE condition ("WHEN") is a fixed value/NULL (e.g. CASE 5+3 > 2 THEN ... ELSE ...)
+   * This avoids computing branches we don't need to compute.
+   */
+  if (when.is_literal()) {
+    const auto when_literal = when.to_literal();
+
+    if (when_literal.value() && !when_literal.null()) {
+      return evaluate_expression<T>(*case_expression.then());
+    } else {
+      return evaluate_expression<T>(*case_expression.else_());
+    }
+  }
+
+  /**
+   * Handle cases where the CASE condition is a series
+   */
+  const auto when_series = when.to_series();
```

Comment:
Ah, I start to get an idea of how you deal with multiple THENs. What is the advantage of introducing the "series" concept compared to "else if"?